Blue green deployments 2 (#4403)
Kong/docs.konghq.com: app/gateway/changelog.md

* The function `kong.db.*.cache_key()` now falls back to `.id` if nothing from `cache_key` is found.
  [#8553](https://github.com/Kong/kong/pull/8553)

### Known limitations

* Kong Manager does not currently support the following features:
  * Secrets management
  * Plugin ordering
  * Expression-based routing

> * Blue-green upgrades from versions of 2.8.1(2.8.1.x for EE) and below to 3.0.0(3.0.0.0 for EE) are not supported for traditional mode.

Reviewer: ```suggestion
* Blue-green upgrades from versions of 2.8.1.x and below to 3.0.0.0 are not supported for traditional mode.
```

* The function `kong.db.*.cache_key()` now falls back to `.id` if nothing from `cache_key` is found.
  [#8553](https://github.com/Kong/kong/pull/8553)

> ### Known limitations

Author: IMO, known limitations should go above breaking changes - thoughts?

* Kong Manager does not currently support the following features:
  * Secrets management
  * Plugin ordering
  * Expression-based routing

* Blue-green deployments:
  * **Kong Gateway (OSS)**: Upgrades with blue-green deloyments are not supported for major versions, therefore they are not supported with upgrades from 2.x.x to 3.0.x.

>   * **Kong Enterprise**: Blue-green upgrades from versions 2.1.x.x-2.7.x.x to 3.0.0.0 are not currently supported.

Reviewer: Well I clearly misunderstood the situation.

Author: Now I'm asking to ensure I have the correct understanding too 😓

Reviewer 2: There are 2 kinds of migration that has no downtime.
1. For a hybrid mode deployment with no version compatibility between versions, we could first upgrade to a new CP, start new DPs, and gracefully stop old DPs. During the process old DPs will not be able to accept new config;
2. For both modes (hybrid and traditional), next 2.8.x.x of EE supports blue/green migration so the old instances can run on the migrated database. But it may cause problems when you change the config with an old instance via admin API for traditional mode.

However, I don't know whether the first kind of migration should be called blue/green.

Reviewer 2: Confirmed that this is the blue/green migration. Going to submit the modification.

>     * This is a known issue planned to be fixed in the next 2.8.x.x release. When that version is released 2.x users should upgrade to that version before beginning a blue-green upgrade to 3.0.0.0.

Reviewer: ```suggestion
    * This is a known issue planned to be fixed in the next 2.8.x.x release. When that version is released, 2.x users should upgrade to that version before beginning a blue-green upgrade to 3.0.0.0.
```

>     * We will remove this known limitation after releasing the next 2.8.x.x version.

Reviewer: ```suggestion
    * We will remove this known limitation with the release of the next 2.8.x.x version.
```

* The function `kong.db.*.cache_key()` now falls back to `.id` if nothing from `cache_key` is found.
  [#8553](https://github.com/Kong/kong/pull/8553)

### Known limitations

* Kong Manager does not currently support the following features:
  * Secrets management
  * Plugin ordering
  * Expression-based routing

* Creating and updating Kong configuration for traditional mode is not allowed during migration from 2.x to 3.0.0.0.
  * This is a known issue planned to be fixed in the next 2.8 release. If this is a requirement for upgrading, Kong operators should upgrade to that version before beginning a upgrade to 3.0.0.0.

>   * In traditional and hybrid mode Kong can be upgraded while continuing to proxy traffic.

Reviewer 2: We are not going to support configuring while migrating. What we are trying to solve with 2.8.2 is we have downtime migrating with traditional mode(proxy port).

> * Creating and updating Kong configuration for hybrid mode is not allowed from both a 2.8 CP and 3.0 CP.

Reviewer 2: It is OK for 3.0 CP to push config to 2.8 DP for EE. We just don't recommend doing that.
And it should not be considered a "known limitation" to not allow different versions of CP to write to DB.

  * Plugin ordering
  * Expression-based routing

> * Blue-green upgrades from versions of 2.8.1(2.8.1.x for EE) and below to 3.0.0(3.0.0.0 for EE) are not supported for traditional mode.

Reviewer 3: As we do not support blue/green upgrades between major versions for CE, there does not seem to be the need to mention CE version numbers here.

Author: https://github.com/Kong/docs.konghq.com/pull/4403#discussion_r966534435

Reviewer 2: > As we do not support blue/green upgrades between major versions for CE, there does not seem to be the need to mention CE version numbers here.

Actually, we are going to support blue-green for 2.8.2 CE
